src/hotspot/cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp

```diff
@@ -779,15 +801,12 @@ void ZBarrierSetAssembler::copy_store_at(MacroAssembler* masm,
 
     // Store barrier pre values and color new values
     if (bytes == 32) {
-      copy_store_barrier(masm, vec_tmp1, src1, Address(dst.base(), dst.offset() + 0), tmp1, tmp2, tmp3, vec_tmp3);
+      copy_store_barrier(masm, vec_tmp1, src1, Address(dst.base(), dst.offset() + 0),  tmp1, tmp2, tmp3, vec_tmp3);
```

Comment:
Since you removed the store below that used to perform the actual store, the copy_store_barrier function kind of needs to perform the store instead, or we will lose one of the stores. But reading that function, it looks like it only performs the store *sometimes*. In the fast path color_and_store_source is used in the fast path, which does indeed perform the store. But when the slow path is used, the "other" copy_store_barrier for normal registers is used, but that one does not perform the store. So when we go through the slow path, the stores are lost and we effectively don't copy as intended.

Now the intended contract which is shared across all GC barrier code across all architectures, is that the barriers never have the responsibility of performing the accesses. This bug demonstrates why - because the responsibilities need to be clear or we get bugs. The store below should be re-instated, and the copy_store_barrier should never perform the actual store.

Reply:
Hi @fisk 

Thanks for your careful review!

x86_64 perform the store in the slow-path:

```
void ZBarrierSetAssembler::copy_store_at(MacroAssembler* masm,
                                         DecoratorSet decorators,
                                         BasicType type,
                                         size_t bytes,
                                         Address dst,
                                         XMMRegister src,
                                         Register tmp1,
                                         Register tmp2,
                                         XMMRegister xmm_tmp) {
...
  } else if (bytes == 32) {
    // 256 bits vector

    if (!dest_uninitialized) {
      Load destination vector
      Check destination store_bad_mask
      __ jcc(Assembler::notZero, fallback);
    }    

 =>   Color and store source // fast-path
    __ jmp(done);

 =>   __ bind(fallback); // slow-path

    __ subptr(rsp, wordSize * 4);
    Store 256 bits vector register in the stack

    // The lower 64 bits.
    Load tmp1 from stack disp 0
=>    copy_store_at dst0 disp 0  // Perform the store in the slow-path

    // The mid-lower 64 bits.
    Load tmp1 from stack disp 8
    copy_store_at dst1 disp 8

    // The mid-higher 64 bits.
    Load tmp1 from stack disp 16
    copy_store_at dst2 disp 16

    // The higher 64 bits.
    Load tmp1 from stack disp 24
    copy_store_at dst3 disp 24

    __ addptr(rsp, wordSize * 4);

    __ bind(done);
  }
...
```

aarch64 also perform the store in the slow-path:

```
static void copy_store_barrier(MacroAssembler* masm,
                               FloatRegister pre_ref,
                               FloatRegister new_ref,
                               Address src, 
                               Register tmp1,
                               Register tmp2,
                               Register tmp3,
                               FloatRegister vec_tmp) {
  Label done, fallback;

  Test reference against bad mask
  __ br(Assembler::NE, fallback);

=>  Color and store source // fast-path
  __ b(done);

=>  __ bind(fallback); // slow-path

  // Extract the 2 oops from the pre_ref vector register
  Unsigned Move vector element The lower 64 bits to general-purpose register
=>  copy_store_barrier(masm, tmp2, noreg, Address(src.base(), src.offset() + 0), tmp1, tmp3); // Perform the store in the slow-path

  Unsigned Move vector element The higher 64 bits to general-purpose register
  copy_store_barrier(masm, tmp2, noreg, Address(src.base(), src.offset() + 8), tmp1, tmp3);

  __ bind(done);
}
```

Please point out my fault if I misunderstood!

> Now the intended contract which is shared across all GC barrier code across all architectures, is that the barriers never have the responsibility of performing the accesses. This bug demonstrates why - because the responsibilities need to be clear or we get bugs.

I will double check again about the `src/hotspot/share/gc/z` code, because jtreg:tier1 and SPECjbb2015 were not able to reproduce such as `non-monotonic marked old transition`, `ZBarrier::mark_slow_path`, etc. assert failtures.

Thanks,
Leslie Zhai